### PR TITLE
Add base conversion lookup table

### LINKED ISSUE #211: keccak: Implement base9_to_base13 conversion gadget
As is mentioned in the specs. Inside of the Mixing step we do the following procedure:
```python
# last round:
    if next_mixing is None:
        # if we don't have new input to mix, just add the round_constant
        state = iota_b9(
            state,
            convert_base(round_constant[23], _from=2, _to=9),
        )
    else:
        # if we have new input to mix, add the round_constant after the base conversion.
        # This is the case the coefficient of `state[0][0]` in the next theta could be 2.
        state = absorb(state, next_mixing)
        state = convert_base(state, _from=9, _to=13)
        state = iota_b13(
            state,
            convert_base(round_constant[23], _from=2, _to=13),
        )
```

`MixingConfig` has been completely implemented in #144 but I realized today that we are missing a gadget that does:
>  state = convert_base(state, _from=9, _to=13)

Indeed, the gadget will be called 24 times during an entire keccak round execution as can be seen in the [specs](https://hackmd.io/sK7v0lr8Txi1bgION1rRpw?view#Overview).

Therefore the gadget should follow the following logic/guidelines:
- Has a `flag` which is assigned to an `Advice` column and copied from another Cell value previously allocated by other Configs which enables or disables the gate.
- This `flag` should only take effect when we are in the `Mixing` step. When we're not, we also call this gadget and we don't need to worry about conditionally enabling constraints. Therefore, we could simply ignore the flag.

## How to do this?
`IotaB9Config` does the same thing (but with a different circuit logic). I highly recommend to take a look to: https://github.com/appliedzkp/zkevm-circuits/blob/3fe115611308f88eccc1c5d61e9a68d1b033e893/keccak256/src/gates/iota_b9.rs#L17-L24

As you can see in the `configure` impl of the struct, you can have to gates that depend on the selector that is enabled (if is or is not in a mixing stage).
Also, note how you can provide two different state assignment functions depending if you're or not in a mixing step:
- https://github.com/appliedzkp/zkevm-circuits/blob/3fe115611308f88eccc1c5d61e9a68d1b033e893/keccak256/src/gates/iota_b9.rs#L105
- https://github.com/appliedzkp/zkevm-circuits/blob/3fe115611308f88eccc1c5d61e9a68d1b033e893/keccak256/src/gates/iota_b9.rs#L80

I basically see the thing as something like:
```rust
#[derive(Clone, Debug)]
pub struct B9ToB13ConvConfig<F> {
    q_not_last: Selector,
    q_last: Selector,
    state: [Column<Advice>; 25],
    _marker: PhantomData<F>,
}

impl<F: FieldExt> B9ToB13ConvConfig<F> {
    // We assume state is recieved in base-9.
    pub fn configure(
        meta: &mut ConstraintSystem<F>,
        state: [Column<Advice>; 25],
    ) -> B9ToB13ConvConfig<F> {
        let q_not_last = meta.selector();
        let q_last = meta.selector();

        // If we're not in the mixing stage, we activate q_not_last and ignore the flag query
        meta.create_gate("b9_to_b13 conversion in steady stage", |meta| {
            let q_enable = meta.query_selector(q_not_last);
            // Gate logic like the lookup call here.
            vec![q_enable.clone() * whatever]
        });

        // If we're in the mixing stage, we activate q_last and mul to it the flag.
        // When flag = false, the gate will be enabled (as is the same with Absorb and IotaB13)
        meta.create_gate("b9_to_b13 conversion in steady state in mixing stage", |meta| {
            let q_enable = {
                let flag = Expression::Constant(F::one()) - meta.query_advice(
                    column_where_you_have_flag,
                    Rotation::whatever(),
                );
                meta.query_selector(q_last) * flag
            };

            // Gate logic like the lookup call here.
            vec![q_enable.clone() * whatever]
        });

        B9ToB13ConvConfig {
            q_not_last,
            q_last,
            state,
        }
    }
```

@therealyingtong I would appreciate that you can double check this. But I think it's in-line with the rest of the implementation.

@ChihChengLiang for any doubts, ping me or we can also keep the discussion updates here. Whatever you prefer :smile:

### PR DESCRIPTION
**Pull request by ChihChengLiang**
Fix #211

## TODO

- slice the lane and convert base chunks by chunks
- glue it to  #144

### CODE REVIEW

File: keccak256/src/gates/state_conversion.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+use crate::gates::tables::BaseInfo;
+use pairing::arithmetic::FieldExt;
+use std::convert::TryInto;
+
+#[derive(Debug, Clone)]
+pub struct StateBaseConversion<F> {
+    bi: BaseInfo<F>,
+    bccs: [BaseConversionConfig<F>; 25],
+    state: [Column<Advice>; 25],
+}
```
Comment (CPerezz) [inline keccak256/src/gates/state_conversion.rs:16]: The base conversion as mentioned in https://github.com/appliedzkp/zkevm-circuits/issues/211 can be used **conditionally** inside of Mixing. 

Therefore, I suggest a similar configuration approach as the one used in `IotaB9` in #144 .

The gate has to be:
- Conditionally enabled due to a flag.
- Be able to be called inside and outside of Mixing (outside the flag is useless).

File: keccak256/src/gates/state_conversion.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            .try_into()
+            .unwrap();
+
+        Self { bi, bccs, state }
+    }
+
+    pub fn assign_region(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        input_state: [F; 25],
```
[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply keccak256/src/gates/state_conversion.rs]: Addressed this one.

**Comment by CPerezz** [comment]
@ChihChengLiang feel free to ask for a new review when is ready! :)

Will wait until you ask.

**Comment by ChihChengLiang** [comment]
sure, I'm still on the flag and the selector. Will ask for a rereview once complete.

File: keccak256/src/gates/state_conversion.rs (hunk 1 above)
Comment (ChihChengLiang) [reply keccak256/src/gates/state_conversion.rs:16]: @CPerezz I added flags on StateBaseConversion and BaseConversionConfig. The flag has these properties:

1. flag is binary
2. parent flag can enable or disable the children's polys.

- The flag is a column, all cells are copied from the parent flag.
- The parent flag is assumed to be constrainted to be 0 or 1. (full fill property 1)
- All polys multiply the flag.  (full fill property 2)

File: keccak256/src/gates/base_conversion.rs [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+            output_coef,
+            output_acc,
+        }
+    }
+
+    pub fn assign_region(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        input: (Cell, F),
+        parent_flag: (Cell, F),
```
Comment (CPerezz) [inline keccak256/src/gates/base_conversion.rs:95]: I'm not sure I would use `parent` here. Just `flag` seems fine already.

File: keccak256/src/gates/base_conversion.rs [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+                        region
+                            .constrain_equal(input_acc_cell, input_coef_cell)?;
+                        region.constrain_equal(
+                            output_acc_cell,
+                            output_coef_cell,
+                        )?;
+                    } else if offset == input_coefs.len() - 1 {
+                        region.constrain_equal(input_acc_cell, input.0)?;
+                        return Ok((output_acc_cell, output_acc));
+                    }
```
Comment (CPerezz) [inline keccak256/src/gates/base_conversion.rs:158]: There's a length missmatch here. 2 constraints in one branch of the `if` and 1 in the `else`. 

Could you confirm if this is an issue @therealyingtong ??

File: keccak256/src/gates/state_conversion.rs (hunk 1 above)
Comment (CPerezz) [reply keccak256/src/gates/state_conversion.rs:16]: Makes complete sense! I like it!!!

File: keccak256/src/gates/tables.rs [hunk 5]
```diff
@@ -188,3 +191,286 @@ impl<F: FieldExt> SpecialChunkTableConfig<F> {
         config
     }
 }
+
+#[derive(Clone, Debug)]
+pub struct BaseInfo<F> {
```
Comment (CPerezz) [inline keccak256/src/gates/tables.rs:196]: Shouldn't this be `pub(crate)`?

File: keccak256/src/gates/tables.rs [hunk 6]
```diff
@@ -188,3 +191,286 @@ impl<F: FieldExt> SpecialChunkTableConfig<F> {
...
+                });
+                F::from(coef)
+            })
+            .collect();
+        Ok((input_coefs, output_coefs, output))
+    }
+}
+
+const MAX_CHUNKS: usize = 64;
+const NUM_OF_BINARY_CHUNKS: usize = 16;
```
Comment (CPerezz) [inline keccak256/src/gates/tables.rs:283]: We should probably move to the top of the File

**Review by CPerezz — COMMENTED** [review]
LGTM. But would like to see @therealyingtong 's opinion on the `if/else` containing constraints.

File: keccak256/src/gates/base_conversion.rs (hunk 4 above)
Comment (ChihChengLiang) [reply keccak256/src/gates/base_conversion.rs:158]: This branches on the offset. Just want a difference on offset 0 and the last offset. 

I think what might matter is how can we be sure the prover and the verifier can drive the same number of rows. The answer is they should have exactly `bi.slice_count()` rows. `input_coefs.len() == output_coefs.len() == bi.slice_count()`. But not sure how to express that here is better.

Comment (therealyingtong) [reply keccak256/src/gates/base_conversion.rs:158]: Branching on the offset looks good to me, since we can deterministically compute that.

**Review by therealyingtong — APPROVED** [review]
LGTM

File: keccak256/src/gates/base_conversion.rs [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+};
+
+use crate::gates::tables::BaseInfo;
+use pairing::arithmetic::FieldExt;
+
+#[derive(Clone, Debug)]
+pub(crate) struct BaseConversionConfig<F> {
+    q_running_sum: Selector,
+    q_lookup: Selector,
+    bi: BaseInfo<F>,
```
[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply keccak256/src/gates/base_conversion.rs]: addressed

**Review by CPerezz — APPROVED** [review]
LGTM! Brillian job @ChihChengLiang :smile:

**Comment by ChihChengLiang** [comment]
Thank you @CPerezz and @therealyingtong for the review!

**Pull request merged**